### ISSUE TITLE
[BUG][Typescript fetch] JSON-serialization does not obay timezone for date fields

### ISSUE THREAD

**Issue by stephanpelikan** [OpenAPITools/openapi-generator#12490]
##### Description

I'm located in Austria which is GMT+1 and in daylight-saving period it is GMT+2. When I have this API definition
```yaml
openapi: 3.0.0
...
components:
  schemas:
    MemberApplicationForm:
      type: object
      properties:
        birthdate:
          type: string
          format: date
```
and set a Javascript Date `new Date('2022-06-05T00:00:00')` then the `toISOString()` method returns `2022-06-04T22:00:00.000Z` which is used as a bases for date formatting in JSON. Therefore the date is '2022-06-04' instead of '2022-06-05'!

Files affected:
1. modules/openapi-generator/src/main/resources/typescript-fetch/apis.mustache
2. modules/openapi-generator/src/main/resources/typescript-fetch/modelGeneric.mustache

##### openapi-generator version

6.0.0

##### Generation Details

Maven build:
```xml
			<plugin>
				<groupId>org.openapitools</groupId>
				<artifactId>openapi-generator-maven-plugin</artifactId>
				<version>6.0.0</version>
				<executions>
					<execution>
						<id>generate-administration-api</id>
						<goals>
							<goal>generate</goal>
						</goals>
						<configuration>
							<inputSpec>${project.basedir}/src/main/resources/api/administration/v1.yaml</inputSpec>
							<output>${project.build.directory}/generated-sources</output>
							<generatorName>spring</generatorName>
							<apiPackage>at.demo.administration.api.v1</apiPackage>
							<modelPackage>at.demo.administration.api.v1</modelPackage>
							<generateSupportingFiles>true</generateSupportingFiles>
							<supportingFilesToGenerate>ApiUtil.java</supportingFilesToGenerate>
							<configOptions>
								<sourceFolder>openapi</sourceFolder>
								<interfaceOnly>true</interfaceOnly>
								<useSpringController>true</useSpringController>
								<dateLibrary>java8</dateLibrary>
								<useBeanValidation>true</useBeanValidation>
							</configOptions>
							<addCompileSourceRoot>true</addCompileSourceRoot>
						</configuration>
					</execution>
				</executions>
			</plugin>
```

##### Steps to reproduce

See description

##### Related issues/PRs

This PR also belongs to date/time in Typescript fetch but does not fix that issue: https://github.com/OpenAPITools/openapi-generator/pull/11685/commits/2cde020568701b92b3a347d26fc8e746c172bbb0#

##### Suggest a fix

```javascript
const offset = yourDate.getTimezoneOffset()
yourDate = new Date(yourDate.getTime() - (offset*60*1000))
return yourDate.toISOString().substr(0,10);
```

**Comment by bjarkehs** [comment]
I created a bug #10224 last year. I think it's an unfortunate approach to use `toISOString` and hacking around it by adding offset to the date.

**Comment by stephanpelikan** [comment]
> I created a bug #10224 last year.

Sorry, I searched for but didn't find it.

>I think it's an unfortunate approach to use `toISOString` and hacking around it by adding offset to the date.

Yes, indead. The main reason for this is that there is no "standard serialization" of dates and times. There are a couple of "toXXXString" methods. If you want to have a specific format one needs to deal with all the details on our own. All that code needs to be packed into a helper-function. I'm not aware of how to introduce such a function in terms of the openapi-generator structure. Besides that I need a fix for this since we rely on OpenAPI and I hope to get a quick fix which can be replaced by a more proper solution later on which may also be applied to other Typescript/Javascript templates.

**Comment by bjarkehs** [comment]
> Sorry, I searched for but didn't find it.

It was also well hidden 😄 

> Yes, indead. The main reason for this is that there is no "standard serialization" of dates and times. There are a couple of "toXXXString" methods. If you want to have a specific format one needs to deal with all the details on our own. All that code needs to be packed into a helper-function. I'm not aware of how to introduce such a function in terms of the openapi-generator structure. Besides that I need a fix for this since we rely on OpenAPI and I hope to get a quick fix which can be replaced by a more proper solution later on which may also be applied to other Typescript/Javascript templates.

I ended up using typescript-axios instead, since it left serialization to me and just returned strings. For me in that project it was better to use strings so I knew exactly what the backend would receive, and I could provide the user's timezone correctly.

**Comment by WizardPC** [comment]
I have the same issue and try to "solve it" by using string instead of Date, but facing a new issue : [12606](https://github.com/OpenAPITools/openapi-generator/issues/12606)

@bjarkehs "do the same" but with switching to axios, that's right ?

I will be interrested if there is another solution :+1:

**Comment by bjarkehs** [comment]
Yeah, I switched to axios for this reason alone. I really wish you could configure typescript-fetch to not handle dates.

Maybe I'll just have to create my own generator that's not trying to be smart about stuff like this.

**Comment by coatesap** [comment]
Now that #11685 is merged, the best solution is probably to provide strings. I can't see that there is going to be any correct way to serialise a `Date` to a date string.

When the [Temporal API](https://tc39.es/proposal-temporal/) lands, we could add `Temporal.PlainDate` as an accepted type, which would be ideal.

**Comment by ymizushi** [comment]
I solve this problem by adding custom class and modifying mustache template like below as last resort.
This solution is not elegent and I strongly hope that this problem will be solved in a smart way.

https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/typescript-fetch/models.index.mustache
```
import { DateTime } from 'src/to/datetime.ts'
export { DateTime } from 'src/to/datetime.ts'

export function DateTimeFromJSON(json: string): DateTime {
  return new DateTime(json);
}

export function DateTimeToJSON(value?: DateTime|null): string|null|undefined {
  if (value === undefined) {
    return undefined;
  if (value === null) {
    return null;
  }
  return value.toString();
}
```

https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/typescript-fetch/runtime.mustache
```typescript
import { DateTime } from 'src/to/datetime.ts'
export { DateTime } from 'src/to/datetime.ts'
```

src/to/datetime.ts
```typescript
// DateTime class implements this interface 
interface DateTimeInterface {
  function toString(): string
  new (): DateTime
}
```

https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/typescript-fetch/modelGeneric.mustache
rename `.toISOString().substr()(0,10)` to `.toString()`

add `--type-mappings=string+date=DateTime` command line option to `openapi-generator-cli`
